### DISCUSSION TITLE
Shim groups

### DISCUSSION THREAD

**Discussion by febrahiminia** [shimming-toolbox/shimming-toolbox#549] [General] [answered]
I am trying to understand the results for the training session as discussed in https://shimming-toolbox.org/en/latest/user_section/tutorials/b0_dynamic_shimming.html#id1. The input fieldmap has 5 slices, and I’m a bit unclear on what the term "shim groups" refers to in this context. In this example, the number of shim groups is 12, and it appears that there are 12 overlapping masks for each slice. I’m wondering why the toolbox doesn’t aggregate these 12 masks into a single mask per slice?

Additionally, I’m running dynamic shimming on the same dataset in Python. While reviewing the shimming evaluation statistics in the log file, I noticed that the standard deviation for the shimmed group is higher than for the unshimmed group, though the overall standard deviation is lower (as shown in the figure fig_shimmed_vs_unshimmed). Could you clarify how I should compare the results for each slice?

Thank you for your help!

**Comment by po09i** [comment, marked as answer]
Hi @Fateme-Nia!

> The input fieldmap has 5 slices, and I’m a bit unclear on what the term "shim groups" refers to in this context.

The term `shim-groups` refers to the number of shim optimizations to perform. In the tutorial, we do a slice-wise shim. Since the image to shim (anat) has 12 slices, there are 12 `shim-groups`. The reason that there are 12 masks is because we are optimizing in the native resolution of the field map. A mask for each `shim-group` optimization is therefore computed which reflects the ROI in the field map space for each slice of the "anat". There is an overlap between each mask of different `shim-groups` controlled by the option "--mask-dilation-kernel-size". This allows to have a good shim in the through slice direction when we do slice-wise shimming.

>  While reviewing the shimming evaluation statistics in the log file, I noticed that the standard deviation for the shimmed group is higher than for the unshimmed group, though the overall standard deviation is lower (as shown in the figure fig_shimmed_vs_unshimmed).

It is possible that a higher STD is computed after an optimization since we are optimizing for the MAE/MSE/RMSE (`--optimizer-criteria`). However, as you point out, we would expect the STD to go down. A reason why the STD could go up is that we are optimizing on a bigger region than the slice itself (as mentioned above [controlled by `--mask-dilation-kernel-size`]), but we are computing the STD only in the slice. It is therefore possible that the STD could go up in the slice but that we would help overall. But, it is not a good sign.

Another reason could be that the optimizer is performing poorly. We have noticed that the `least_squares` optimizer sometimes performs poorly when the ROI is small. I recommend using the `pseudo_inverse` method if you are not limited by the bounds of the coil. We are looking into this small ROI issue for the `least_squares` method with MSE. I have noticed that using `--optimizer-criteria rmse` can help. I have not tested this extensively, but in the tutorial for example, using RMSE gives a better STD.

Since the STD can be better with other optimizers, I suspect the issue comes from the `least_squares` optimizer and MSE criteria combination and the small ROI.

Conclusion:
- If you are not current limited, use `pseudo_inverse`
- If you are current limited, use `least_squares` with `--optimizer-criteria rmse` and hopefully that helps
- I will see if I can find an issue with MSE, I might put RMSE as the default option.

Let me know if you have other questions!
Alex
